hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bloom/HoodieBloomMetaIndexLazyCheckFunction.java

```diff
@@ -0,0 +1,137 @@
+import org.apache.log4j.Logger;
+import org.apache.spark.api.java.function.Function2;
+import scala.Tuple2;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Spark Function2 implementation for checking bloom filters for the
+ * requested keys from the metadata table index. The bloom filter
+ * checking for keys and the actual file verification for the
+ * candidate keys is done in a lazy fashion of file by file.
+ */
+public class HoodieBloomMetaIndexLazyCheckFunction implements Function2<Integer, Iterator<Tuple2<String, HoodieKey>>, Iterator<List<HoodieKeyLookupResult>>> {
+
+  private static final Logger LOG = LogManager.getLogger(HoodieBloomMetaIndexLazyCheckFunction.class);
+  private final HoodieTable hoodieTable;
+  private final HoodieWriteConfig config;
+
+  public HoodieBloomMetaIndexLazyCheckFunction(HoodieTable hoodieTable, HoodieWriteConfig config) {
+    this.hoodieTable = hoodieTable;
+    this.config = config;
+  }
+
+  @Override
+  public Iterator<List<HoodieKeyLookupResult>> call(Integer integer, Iterator<Tuple2<String, HoodieKey>> tuple2Iterator) throws Exception {
+    return new LazyKeyCheckIterator(tuple2Iterator);
+  }
+
+  class LazyKeyCheckIterator extends LazyIterableIterator<Tuple2<String, HoodieKey>, List<HoodieKeyLookupResult>> {
+
+    private HoodieKeyLookupHandle keyLookupHandle;
+
+    LazyKeyCheckIterator(Iterator<Tuple2<String, HoodieKey>> filePartitionRecordKeyTripletItr) {
+      super(filePartitionRecordKeyTripletItr);
+    }
+
+    @Override
+    protected void start() {
+    }
+
+    @Override
+    protected List<HoodieKeyLookupResult> computeNext() {
+
+      List<HoodieKeyLookupResult> ret = new ArrayList<>();
+      try {
+        final Map<String, HoodieBaseFile> fileIDBaseFileMap = new HashMap<>();
+        while (inputItr.hasNext()) {
+          Tuple2<String, HoodieKey> currentTuple = inputItr.next();
+          final String partitionPath = currentTuple._2.getPartitionPath();
+          final String recordKey = currentTuple._2.getRecordKey();
+          final String fileId = currentTuple._1;
+          ValidationUtils.checkState(!fileId.isEmpty());
+          if (!fileIDBaseFileMap.containsKey(fileId)) {
+            Option<HoodieBaseFile> baseFile = hoodieTable.getBaseFileOnlyView().getLatestBaseFile(partitionPath, fileId);
+            if (!baseFile.isPresent()) {
+              throw new HoodieIndexException("Failed to find the base file for partition: " + partitionPath
+                  + ", fileId: " + fileId);
+            }
+            fileIDBaseFileMap.put(fileId, baseFile.get());
+          }
+          final Pair<String, String> partitionPathFileIdPair = Pair.of(partitionPath, fileId);
+
+          // lazily init state
+          if (keyLookupHandle == null) {
+            keyLookupHandle = new HoodieKeyLookupHandle(config, hoodieTable, partitionPathFileIdPair,
```

Comment:
I think the issue here is that we open the metadata table again and again each time. Why can't we open the bloom partition once here and pass it into the handle? That way we only incur the merge cost once?

Reply:
File by File bloom check is no more used. Consolidated the bloom filter check into batch mode done in iterable fashion.